### PR TITLE
Feature/#2 default implementation of IObject

### PR DESCRIPTION
**Pull request by omskscream**
Serialization was implemented in ```toString()``` method

### CODE REVIEW

File: src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java [hunk 1]
```diff
@@ -47,4 +48,14 @@ public void setValue(FieldName name, Object newValue) throws ChangeValueExceptio
         }
         map.put(name.toString(), newValue);
     }
+
+    @Override
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: As we disscussed before it's not an acceptably.

File: src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java [hunk 2]
```diff
@@ -47,4 +48,14 @@ public void setValue(FieldName name, Object newValue) throws ChangeValueExceptio
...
+    public String toString() {
+        try {
+            return objectMapper.writeValueAsString(map);
+        } catch (JsonProcessingException e) {
+            // this method must not throw any exceptions
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: As we discussed before this is unacceptably

File: src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java [hunk 3]
```diff
@@ -18,13 +19,13 @@
 public class SMObject implements IObject {
 
     private Map<String, Object> map;
+    private ObjectMapper objectMapper = new ObjectMapper();
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: Why do you have this object a whole life o the object?

Comment (omskscream) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: Because creation of ObjectMapper costs more than saving it. I also suggest to make this field `static final` according to http://stackoverflow.com/a/3909846

Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: It looks like as a pessimistic optimization. Also how do you calclate it?

File: src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java (hunk 2 above)
Comment (omskscream) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: The JsonProcessingException on `writeValueAsString()` was declared to achieve backward compatibility http://jira.codehaus.org/browse/JACKSON-758 and should never be thrown.

Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: Great. But If we catch an JsonProcessingException we should throw own Exception in this case. It looks like we must throw something likes Error exception in this case. In other words we should not have any empty catch block of try-catch. There are no warranty that this exception will came again in a future.

Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: If author change contract we get a compiller error. But if he return back this an exception we will not know about it. So if exception is in contract we must handle it!

Comment (omskscream) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: I propose to use new java 8 `UncheckedIOException`. It will be correct because `JsonProcessingException` is a wrapper for `IOException` and `UncheckedIOException` 
> Wraps an {@link IOException} with an unchecked exception.
according to its javadoc.

Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: Look this
catch (JsonProcessingException e) {
+            // this method must not throw any exceptions <- THIS IS UNACCEPTABLE
}

Comment (omskscream) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: I meant the following:
```java
public String toString() {
        try {
            return objectMapper.writeValueAsString(map);
        } catch (JsonProcessingException e) {
            throw new UncheckedIOException(e);
        }
    }
```

Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: Yes, please implement what do mean in source code.

File: src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java (hunk 3 above)
Comment (omskscream) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: I used immutable `ObjectReader` and `ObjectWriter` instanses for to-from json manipulations. This approach is fully thread-safe.
http://wiki.fasterxml.com/ObjectReader
http://wiki.fasterxml.com/ObjectWriter

File: src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java [hunk 4]
```diff
@@ -18,15 +20,15 @@
...
 
     public SMObject() {
         this.map = new HashMap<>(0);
     }
 
     public SMObject(String raw) throws IllegalArgumentException {
-        ObjectMapper objectMapper = new ObjectMapper();
         try {
-            this.map = objectMapper.readValue(raw, new TypeReference<Map<String, Object>>() {});
+            this.map = objectMapper.reader(new TypeReference<Map<String, Object>>() {}).readValue(raw);
```
Comment (etyumentcev) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: Has any reson to change this string?

File: src/test/java/info/smart_tools/smartactors/core/IObjectTest.java [hunk 5]
```diff
@@ -1,8 +1,12 @@
 package info.smart_tools.smartactors.core;
```
Comment (etyumentcev) [inline src/test/java/info/smart_tools/smartactors/core/IObjectTest.java]: Please add more tests: get inner object, get empty object, get array, get empty array, get different built-in types, get array of objects

File: src/test/java/info/smart_tools/smartactors/core/IObjectTest.java [hunk 6]
```diff
@@ -68,4 +72,43 @@ public void Should_ThrowExceptionFromConstructorOrFromGetValue_When_Initializati
...
+        String stringVal = impl.toString();
+        assertNotNull(stringVal);
+        assertTrue(emptyJsonPattern.matcher(stringVal).matches());
+    }
+
+    @Test(dataProvider = "Should_ReturnJsonEquivalentToInitial_When_ToStringCalledOnNonEmptyIObject",
+            dataProviderClass = DataProviderForIObjectTest.class)
+    public void Should_ReturnJsonEquivalentToInitial_When_ToStringCalledOnNonEmptyIObject(IObjectCreator creator) {
+        String initialJson = "{\"a\":{\"b\":{\"c\":\"test\"}}}";
+        String initialPrettyJson = "{\n" +
```
Comment (etyumentcev) [inline src/test/java/info/smart_tools/smartactors/core/IObjectTest.java]: Please add to test example: empty array, array of integers, array of objects, empty object, all built-in types

File: src/test/java/info/smart_tools/smartactors/core/IObjectTest.java (hunk 5 above)
Comment (etyumentcev) [inline src/test/java/info/smart_tools/smartactors/core/IObjectTest.java]: Same for write values to object

File: src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java (hunk 4 above)
Comment (omskscream) [inline src/main/java/info/smart_tools/smartactors/core/impl/SMObject.java]: I changed this line after I found out about immutable `ObjectReader` and `ObjectWriter`, which are fully thread-safe. The most recent Jackson documentation recommends to use that classes for json manipulations instead of `ObjectMapper`.
http://wiki.fasterxml.com/ObjectReader
http://wiki.fasterxml.com/ObjectWriter
